compiler/rustc_resolve/src/macros.rs

```diff
@@ -544,11 +544,23 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         if let Some((article, expected)) = unexpected_res {
             let path_str = pprust::path_to_string(path);
             let msg = format!("expected {}, found {} `{}`", expected, res.descr(), path_str);
-            self.tcx
-                .sess
-                .struct_span_err(path.span, &msg)
-                .span_label(path.span, format!("not {} {}", article, expected))
-                .emit();
+            let mut err = self.tcx.sess.struct_span_err(path.span, &msg);
+
+            err.span_label(path.span, format!("not {} {}", article, expected));
+
+            if kind == MacroKind::Derive && ext.macro_kind() != MacroKind::Derive {
+                // Suggest removing the derive() as the macro isn't Derive
+                let opening_span =
+                    path.span.shrink_to_lo().with_lo(path.span.lo() - rustc_span::BytePos(7));
```

Comment:
This may not work when:
- there is space between the derive and the path? `#[ derive( Path ) ]`
- there are multiple paths in the same derive? `#[derive(A, B)]`
- the path comes from a macro expansion? `#[derive($path)]` in a macro with `$path` interpolation variable.

Reply:
Yes, this was part of my issues, what is the proper way to get the parent `Span`?

Reply:
I wonder if it would be better to:

- Skip this if the `Span` is from a macro, because it's probably difficult to make an appropriate suggestion in that context.
- Suggest removing the `Path` from `#[derive(Path)]` and suggest adding a new line with `#[Path]` instead?
  - This would be better for the `#[derive(A, B)]` case but worse for the `#[derive(A)]` case.

Reply:
Yes I think that's reasonable - after a lot of poking around I think it's unreasonable to get the parent span as it would require some significant changes (far more than I am comfortable with currently)

Reply:
Could you clarify what you mean by:
> Skip this if the Span is from a macro, because it's probably difficult to make an appropriate suggestion in that context.

Do you mean as span such as this? As far as I know it isn't valid
```
#[derive(foo!())]
         ^^^^^^
```

Reply:
I just mean using [`Span::from_expansion`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.from_expansion) to check whether this code was generated by a macro, since there's not much we can do with `#[derive($foo)]`.

Reply:
Oh I see - many thanks